### PR TITLE
KAFKA-19248: Multiversioning in Kafka Connect - Plugin Loading Isolation Tests

### PR DESCRIPTION
**Pull request by snehashisp**
This adds tests for
[KIP-891](https://cwiki.apache.org/confluence/display/KAFKA/KIP-891%3A+Running+multiple+versions+of+Connector+plugins).
It primarily focuses on tests for the new additions in plugin loading
isolation. It has dependency on the actual KIP implementation PRs and
should be merged post https://github.com/apache/kafka/pull/17742

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+                Assertions.assertInstanceOf(PluginClassLoader.class, pluginLoader);
+                Assertions.assertTrue(((PluginClassLoader) pluginLoader).location().contains(pluginLocation));
+                Object p = plugins.newPlugin(buildInfo.plugin().className(), PluginUtils.connectorVersionRequirement(buildInfo.version()));
+                Assertions.assertInstanceOf(Versioned.class, p);
+                Assertions.assertEquals(buildInfo.version(), ((Versioned) p).version());
+            }
+        }
+    }
+
+    private static final PluginType[] ALL_PLUGIN_TYPES = new PluginType[]{
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java]: nit: The tests still pass if this variable is set to PluginType.values(). Should they fail? Can we eliminate this variable?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java [hunk 2]
```diff
@@ -477,6 +480,27 @@ private static void compileJavaSources(Path sourceDir, Path binDir) throws IOExc
...
+    private static File copyAndReplace(File source, Map<String, String> replacements) throws RuntimeException {
+        if (replacements.isEmpty()) {
+            return source;
+        }
+        try {
+            String content = Files.readString(source.toPath());
+            for (Map.Entry<String, String> entry : replacements.entrySet()) {
+                content = content.replace(entry.getKey(), entry.getValue());
+            }
+            File tmpFile = new File(System.getProperty("java.io.tmpdir") + File.separator + source.getName());
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:489]: nit: use Files.createTempFile?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java [hunk 3]
```diff
@@ -477,6 +480,27 @@ private static void compileJavaSources(Path sourceDir, Path binDir) throws IOExc
...
+    }
+
+    private static File copyAndReplace(File source, Map<String, String> replacements) throws RuntimeException {
+        if (replacements.isEmpty()) {
+            return source;
+        }
+        try {
+            String content = Files.readString(source.toPath());
+            for (Map.Entry<String, String> entry : replacements.entrySet()) {
+                content = content.replace(entry.getKey(), entry.getValue());
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:487]: nit: I don't really like this "modifying arbitrary source code before compilation" technique, it's very powerful and open-ended, and could be misused and be more difficult to debug like self-modifying code. But because the current application is within reason, maybe this is good enough to merge.

Prior art here is the ReadVersionFromResourcePlugin, whose code is duplicated in read-version-from-resource-v1 and read-version-from-resource-v2, with version files are specified explicitly. I think that we shouldn't follow this same strategy for these new plugins and versions, but there's a middle ground where we generate version files from Strings specified in createPluginJar/writeJar.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java [hunk 4]
```diff
@@ -52,6 +52,7 @@
 import javax.tools.StandardJavaFileManager;
 import javax.tools.ToolProvider;
 
+
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java]: nit: newline

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java [hunk 5]
```diff
@@ -477,6 +480,27 @@ private static void compileJavaSources(Path sourceDir, Path binDir) throws IOExc
...
+            if (!replacements.isEmpty()) {
+                sourceFiles.forEach(File::delete);
+            }
+        }
+    }
+
+    private static File copyAndReplace(File source, Map<String, String> replacements) throws RuntimeException {
+        if (replacements.isEmpty()) {
+            return source;
+        }
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java]: Having the same condition in multiple disconnected places is a common source of bugs. If these two pieces of code are not called together, or someone modifies one condition but not the other, things won't behave correctly.

Could the control flow be simplified? It's worth considering whether you need either of these conditions in the code at all.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class VersionedPluginBuilder {
+
+    private static final String VERSION_PLACEHOLDER = "PLACEHOLDER_FOR_VERSION";
+
+    public enum VersionedTestPlugin {
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java:31]: WDYT about rolling these into the TestPlugin enum?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+    public VersionedPluginBuilder include(VersionedTestPlugin plugin, String version) {
+        pluginBuilds.add(new BuildInfo(plugin, version));
+        return this;
+    }
+
+    public synchronized Path build(String pluginDir) throws IOException {
+        Path pluginDirPath = Files.createTempDirectory(pluginDir);
+        Path subDir = Files.createDirectory(pluginDirPath.resolve("lib"));
+        for (BuildInfo buildInfo : pluginBuilds) {
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java:101]: Allowing the code to combine multiple test plugins together into a single plugin location is very interesting. Currently the bad-packaging plugin relies on plugins being located together, but this is because the code is laid out as multiple source files inside of a single plugin.

File: connect/runtime/src/test/resources/test-plugins/sampling-connector/test/plugins/VersionedSamplingSinkConnector.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * VersionedSamplingSinkConnector is a test connector that extends SamplingConnector and overrides the version method.
+ * Any instance of the string PLACEHOLDER_FOR_VERSION will be replaced with the actual version during plugin compilation.
+ */
+public final class VersionedSamplingSinkConnector extends SamplingConnector {
```
Comment (gharris1727) [inline connect/runtime/src/test/resources/test-plugins/sampling-connector/test/plugins/VersionedSamplingSinkConnector.java]: Is it significant that this is a SamplingConnector or SamplingPlugin, or is that just for convenience? I see that the SamplingPlugin interface is only used in PluginsTest, which doesn't use this (or the other Versioned) plugins.

The SamplingPlugin was meant specifically for the plugins to exfiltrate ClassLoader references to assertions in PluginsTest code, it's not needed if you're not performing those assertions, as you can just extend the connect-api class.

Theoretically you could add classloader assertions to the multiversion tests, but that doesn't feel very high priority, as the existing classloader assertions should have pretty good coverage. Or because Connectors are always Versioned, maybe you could add the version substitution to the SamplingConnector itself, rather than extending it.

One problem with combining both fixed and dynamic versioned plugins together is that you are obligated to bring an additional copy of the fixed version plugin with each dynamic versioned plugin. For example, if you want a 1.0 and 2.0 VersionedSamplingSinkConnector in the same plugin location, you are forced to have two .class files for the 1.0.0 SamplingConnector. This is not generally a supported condition because of classloader nondeterminism, but because it's all compiled from the same source it probably won't break. Unless the code substitution mechanism causes a binary incompatibility....

TLDR: I think the Versioned plugins should not implement SamplingPlugin, and should be in their own packages like the versioned-predicate and versioned-transformation.

This applies to all of the Sampling test plugins.

File: connect/runtime/src/test/resources/test-plugins/sampling-converter/test/plugins/VersionedSamplingConverter.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaAndValue;
+import org.apache.kafka.connect.storage.Converter;
+import org.apache.kafka.connect.runtime.isolation.SamplingTestPlugin;
+
+/**
+ * Converter to test multiverioning of plugins.
+ * Any instance of the string PLACEHOLDER_FOR_VERSION will be replaced with the actual version during plugin compilation.
+ */
+public class VersionedSamplingConverter extends SamplingConverter implements Versioned {
```
Comment (gharris1727) [inline connect/runtime/src/test/resources/test-plugins/sampling-converter/test/plugins/VersionedSamplingConverter.java]: There is definitely value in having a distinct class here, because the Converter interface doesn't extend Versioned.

File: connect/runtime/src/test/resources/test-plugins/versioned-predicate/test/plugins/VersionedPredicate.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.kafka.connect.transforms.predicates.Predicate;
+
+import java.util.Map;
+
+/**
+ /**
+ * Predicate to test multiverioning of plugins.
+ * Any instance of the string PLACEHOLDER_FOR_VERSION will be replaced with the actual version during plugin compilation.
+ */
+public class VersionedPredicate<R extends ConnectRecord<R>> implements Predicate<R>, Versioned {
```
Comment (gharris1727) [inline connect/runtime/src/test/resources/test-plugins/versioned-predicate/test/plugins/VersionedPredicate.java:33]: nit: We have NonMigratedPredicate already, maybe we need a migrated, but non-versioned predicate?
Similar for Transformation.

This is already missing test coverage on trunk, so maybe it can be addressed in a follow-up. I think it's also a bug that you fixed in an earlier PR.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.SINK_CONNECTOR, k -> "0.1.0"));
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.CONVERTER,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.CONVERTER, k -> "0.2.0"));
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.HEADER_CONVERTER,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.HEADER_CONVERTER, k -> "0.3.0"));
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.TRANSFORMATION,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.TRANSFORMATION, k -> "0.4.0"));
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.PREDICATE,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.PREDICATE, k -> "0.5.0"));
+            DEFAULT_COMBINED_ARTIFACT = Collections.singletonMap(builder.build("all_versioned_artifact"), builder.buildInfos());
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java:133]: nit: Why is this computeIfAbsent, when we are constructing the map for the first time?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        Assertions.assertInstanceOf(PluginClassLoader.class, connectorLoader);
+
+        List<VersionedPluginBuilder.VersionedTestPlugin> pluginTypes = List.of(
+            VersionedPluginBuilder.VersionedTestPlugin.CONVERTER,
+            VersionedPluginBuilder.VersionedTestPlugin.HEADER_CONVERTER,
+            VersionedPluginBuilder.VersionedTestPlugin.TRANSFORMATION,
+            VersionedPluginBuilder.VersionedTestPlugin.PREDICATE
+        );
+        // should match the version used in setUp for creating the combined artifact
+        List<String> versions = Arrays.asList("0.2.0", "0.3.0", "0.4.0", "0.5.0");
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java]: Do these versions correspond to the versions specified in the static initializer? Can they be shared constants?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+            artifacts.putAll(DEFAULT_ISOLATED_ARTIFACTS);
+            MULTI_VERSION_PLUGINS = setUpPlugins(artifacts, PluginDiscoveryMode.SERVICE_LOAD);
+
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    public void TestVersionedPluginLoaded() throws InvalidVersionSpecificationException, ClassNotFoundException {
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java]: nit: lowercase the first letter of all of the tests. Here and all tests in PluginRecommenderTest.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java [hunk 14]
```diff
@@ -477,6 +480,27 @@ private static void compileJavaSources(Path sourceDir, Path binDir) throws IOExc
...
+        try {
+            String content = Files.readString(source.toPath());
+            for (Map.Entry<String, String> entry : replacements.entrySet()) {
+                content = content.replace(entry.getKey(), entry.getValue());
+            }
+            File tmpFile = new File(System.getProperty("java.io.tmpdir") + File.separator + source.getName());
+            Files.writeString(tmpFile.toPath(), content);
+            return tmpFile;
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:492]: Please also call deleteOnExit() for this file.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        pluginBuilds = new ArrayList<>();
+    }
+
+    public VersionedPluginBuilder include(VersionedTestPlugin plugin, String version) {
+        pluginBuilds.add(new BuildInfo(plugin, version));
+        return this;
+    }
+
+    public synchronized Path build(String pluginDir) throws IOException {
+        Path pluginDirPath = Files.createTempDirectory(pluginDir);
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java:97]: This method is leaking files on each run. I found a bunch of files left over on my machine:
```
cd /var/folders/k1/6b93_bm16596yf4d3dlp6vbh0000gn/T
ls
rm -rf SINK_CONNECTOR-* SOURCE_CONNECTOR-* PREDICATE-* TRANSFORMATION-* HEADER_CONVERTER-* CONVERTER-* all_versioned_artifact*
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginRecommenderTest.java [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.kafka.connect.runtime.isolation.MultiVersionTest.DEFAULT_COMBINED_ARTIFACT_VERSIONS;
+import static org.apache.kafka.connect.runtime.isolation.MultiVersionTest.DEFAULT_ISOLATED_ARTIFACTS;
+import static org.apache.kafka.connect.runtime.isolation.MultiVersionTest.MULTI_VERSION_PLUGINS;
+
+public class PluginRecommenderTest {
+
+    private Set<String> allVersionsOff(String classOrAlias) {
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginRecommenderTest.java]: nit: off -> of typo?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java (hunk 2 above)
Comment (snehashisp) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:489]: Tried this but it unfortunately always adds a random numerical id to the file generated, even with both prefix and suffix (with one of them as `""`). It will add the id in between the prefix and suffix. This fails compilation as the file name of the class is required to have the same name as the class with the `.java` suffix.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java (hunk 11 above)
Comment (snehashisp) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java:133]: So `computeIfAbsent` conveniently returns the computed value, which is used in the builder. A simple `put` would return the previous value which would be `null`. We can separate out this into individual `put` and 'get` calls if this looks a bit hard to read.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java (hunk 15 above)
Comment (snehashisp) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java:97]: Thanks for catching this. Added deleteOnExit on the created temp plugins so that it gets cleaned up.

File: connect/runtime/src/test/resources/test-plugins/sampling-connector/test/plugins/VersionedSamplingSinkConnector.java (hunk 8 above)
Comment (snehashisp) [reply connect/runtime/src/test/resources/test-plugins/sampling-connector/test/plugins/VersionedSamplingSinkConnector.java]: Makes sense to separate them out. When I wrote this, I was not sure if the sampling plugin helpers would be useful in the new multi versioning test but I did not really need to use them in the end.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java (hunk 5 above)
Comment (snehashisp) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java]: Updated this. When replacements are present we needed to clean up the temporary java files with the replacements, hence this code flow. But with the suggestion of using `deleteOnExit` this check is no longer required.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java (hunk 7 above)
Comment (snehashisp) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java:101]: Yeah, not sure if the need was there before. With multi versioning I wanted to have a test that would correctly use a bundled version of a plugin instead of a latest one, if no version is specified in the config. For instance a converter or transformation that is part of the connector libraries.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java (hunk 6 above)
Comment (snehashisp) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginBuilder.java:31]: The reason I separated them out is because the plugins in TestPlugin are all static and hence are precompiled, whereas for versioned plugins we can generate plugins dynamically with different versions.

File: connect/runtime/src/test/resources/test-plugins/versioned-predicate/test/plugins/VersionedPredicate.java (hunk 10 above)
Comment (snehashisp) [reply connect/runtime/src/test/resources/test-plugins/versioned-predicate/test/plugins/VersionedPredicate.java:33]: The new VersionedTransformation has a service loader manifest, so the new tests should address some of the missing test coverage.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java (hunk 3 above)
Comment (snehashisp) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/TestPlugins.java:487]: Yeah its powerful. It would be interesting to see in what ways people start misusing this capability. If it turns out to be problematic,  I will followup and see if the version file approach can be introduced instead.

**Comment by snehashisp** [comment]
Hi @gharris1727. Have addressed the comments in this PR, PTAL soon.

**Review by gharris1727 — APPROVED** [review]
LGTM.

I was having some difficulty running this locally due to https://issues.apache.org/jira/browse/KAFKA-18834 . As it's unrelated and seems to be well understood, I'll just keep an eye out to see if this causes more problems on trunk.
